packages/babel-traverse/src/scope/index.ts

```diff
@@ -823,10 +830,21 @@ export default class Scope {
       if (!binding) return false;
       if (constantsOnly) return binding.constant;
       return true;
+    } else if (
+      isThisExpression(node) ||
+      isMetaProperty(node) ||
+      isTopicReference(node)
+    ) {
+      return true;
     } else if (isClass(node)) {
       if (node.superClass && !this.isPure(node.superClass, constantsOnly)) {
         return false;
       }
+      if (node.decorators) {
+        for (const decorator of node.decorators) {
+          if (!this.isPure(decorator, constantsOnly)) return false;
```

Comment:
If there is a decorator, it's never pure.

i.e.
```js
const dec = () => console.log(1);

@dec
class A {}
```

here `@dec` is pure because `dec` is a constant, but then it's evaluated.